lib/Factory/MenuFactory.php

```diff
@@ -4,88 +4,147 @@
 
 use Timber\CoreInterface;
 use Timber\Menu;
-
 use WP_Term;
 
 /**
  * Internal API class for instantiating Menus
  */
 class MenuFactory {
-	public function from($params, array $args = []) {
-		if ($params === 0) {
-			// We know no nav_menu term exists with ID 0,
-			// so just look at the existing terms in the database.
+
+	/**
+	 * Try to get a menu by all means available in an order that matches the
+	 * most common use cases. It will fallback on the first menu found if no parameters are provided
+	 * or if no menu is found with the current parameters.
+	 *
+	 * Note that this method has pitfalls and might not be the most performant way to get a menu
+	 *
+	 * @param mixed $params
+	 * @param array $args
+	 * @return Menu|null
+	 */
+	public function from($params, array $args = []) : ?Menu {
+		$menu = null;
+
+		if(empty($params)) {
 			return $this->from_nav_menu_terms($args);
 		}
 
+		// If $params is a numeric slug, we might get the wrong menu
 		if (is_numeric($params)) {
-			return $this->from_id((int) $params, $args);
+			$menu = $this->from_id((int) $params, $args);
 		}
 
-		if (is_string($params)) {
-			return $this->from_string($params, $args);
+		if (is_object($params)) {
+			$menu = $this->from_object($params, $args);
 		}
 
-		if (is_object($params)) {
-			return $this->from_term_object($params, $args);
+		if (!$menu && is_string($params)) {
+			// If $location is the same than some menu slug, we might get the wrong menu
+			$menu = $this->from_location($params, $args);
+			if(!$menu) {
+				$menu = $this->from_slug($params, $args);
+			}
+			if(!$menu) {
+				$menu = $this->from_name($params, $args);
+			}
 		}
 
-		// Fall back on the first nav_menu term we find.
-		return $this->from_nav_menu_terms($args);
+		if(!$menu) {
```

Comment:
While working on the Pages Menu functionality (#2562) I stumbled over this again. And I propose to remove this if-statement.

If we leave this out and just return `null` when no menu couldn’t be determined up until here, then we wouldn’t run into unexpected behaviors where the wrong menu is returned (see #1440).

WordPress uses the following functionality in `wp_nav_menu()`:

```php
// Get the first menu that has items if we still can't find a menu.
if ( ! $menu && ! $args->theme_location ) {
    $menus = wp_get_nav_menus();
    foreach ( $menus as $menu_maybe ) {
        $menu_items = wp_get_nav_menu_items( $menu_maybe->term_id, array( 'update_post_term_cache' => false ) );
        if ( $menu_items ) {
            $menu = $menu_maybe;
            break;
        }
    }
}
```

We replicate this functionality in `MenuFactory::from_nav_menu_terms()`. This is fine if no parameter is passed to `Timber::get_menu()` by choice. However, if a parameter is passed, I think we should return `null` if no menu is found. This would be what I expect the function to do.

Since we have a lot of custom menu logic, I think we don’t have to follow what WordPress does here.

Reply:
Yes, I fully agree! I will update this.

Reply:
I think now the logic is still the same, just switched around.

I propose that we don’t call `$this->from_nav_menu_terms($args)` in `MenuFactory::from()` if no menu can be found. With the current logic, if no menu could be found with the provided args, then Timber will *still* try to return the first menu it can find.

So instead of …

```php
if($menu) {
    return $menu;
}

return $this->from_nav_menu_terms($args);
```

at the end of the method, I would just do …

```php
return $menu;
```

This would fix the bug in #1440 and would be more in line with what I would expect the function to do.

Reply:
Sorry, this was not unfinished, this is now fixed!